### ISSUE TITLE
Forcing layout?

### ISSUE THREAD

**Issue by ELLIOTTCABLE** [jeantil/autokbisw#17]
Soooo, I can't figure out why, but having this installed … it looks like I can't *change* my layout, for a given input device, ever again?

I have autokbisw installed, and registered with LaunchD. I have an external, Colemak-layout mechanical keyboard, that I use at work; autokbisw keeps my macOS layout set to `U.S.` while using my hardware-Colemak board. I -also- use my work laptop on the go; when typing on my internal keyboard, autokbisw sets my macOS layout to `Colemak`.

Problem is, today, after a couple weeks of happily using this setup, I tried to install and run a game on this laptop. I switch to `U.S.` keyboard layout … open the game … and try to WASD around. No dice. I tab back out, and … the macOS layout is back in `Colemak`, even though I manually changed it? I try switching it to `U.S.` (success), ⌘␣ to try it out … and watch the macOS input-source in my menubar immediately switch back to `Colemak` as soon as I touch my keyboard.

**tl;dr** Am I missing something? I thought autokbisw was supposed to only change my layout when I used a *different* keyboard; not keep switching it while I'm on a *single* keyboard, even when I try to manually override it.

**Comment by jeantil** [comment]
This sounds really weird, autokbisw basically is a Map where the keys are keyboard identifiers and the values are layouts. 

It then monitors all lowlevel input events and compares keyboard id of the event with the last seen id, if the id changed, it will switch to the layout associated for this id in the map.

It also monitors layout changes and updates the corresponding entry in the map for the current keyboard id. 

Can you tell me which options you use to run autokbisw ? (in particular are you using the -l flag ?) 

What you describe definitely should not happen can you run stop the daemon, run autokbisw manually with -v flag
Alternatively you can also change your laucnhd plist file to enable output to a log file and add the verbose mode. 

Once this is done, reproduce the issue and please send me the log file.

Thanks!

**Comment by ELLIOTTCABLE** [comment]
I … cannot reproduce the issue. For whatever reason, I assumed it was intended behaviour! If I'd realized it was a bug, I'd have put a little more effort into debugging it. 😱

I'm gonna close this, then, unless the behaviour rears its head again.

**Comment by jeantil** [comment]
Thank you.

If you do encounter this issue again please activate logging and file the trace.

**Comment by ELLIOTTCABLE** [comment]
The problem definitely consistently reproduces; it started happening again this weekend.

However, at the moment, I'm having trouble debugging with `-v` - it doesn't seem to work at all? All keyboards simply stay on whatever input-source I last switched to, as if it weren't running at all:

```
❯ ~ [4s] /usr/local/bin/autokbisw -v
starting with useLocation:false - verbosity:1
set input source to <TSMInputSource 0x7f90a3f0dae0> KB Layout: U.S. (id=0) for keyboard Optional(Karabiner VirtualHIDKeyboard)-[Optional(5824)-Optional(10203)-Optional(pqrs.org)-Optional(pqrs.org:org_pqrs_driver_Karabiner_VirtualHIDDevice_VirtualHIDKeyboard_v060800)]
```

**Comment by jeantil** [comment]
Hi @ELLIOTTCABLE 

the trace has 2 levels (I know it's not obvious) 

The first level only tracks layout changes triggered by autokbisw. for example 
```
starting with useLocation:false - verbosity:1
set input source to <TSMInputSource 0x7fe5bb61a2e0> KB Layout: French (id=1) for keyboard Optional(Apple Internal Keyboard / Trackpad)-[Optional(1452)-Optional(628)-Optional(Apple Inc.)-Optional(D3H51914381FTV4AQ2BS)]
set input source to <TSMInputSource 0x7fe5bb61a2e0> KB Layout: French (id=1) for keyboard Optional(Dell USB Keyboard)-[Optional(16700)-Optional(8208)-Optional(Dell)-nil]
set input source to <TSMInputSource 0x7fe5bb61a2e0> KB Layout: French (id=1) for keyboard Optional(Apple Internal Keyboard / Trackpad)-[Optional(1452)-Optional(628)-Optional(Apple Inc.)-Optional(D3H51914381FTV4AQ2BS)]
set input source to <TSMInputSource 0x7fe5bb40a680> KB Layout: Colemak (id=12825) for keyboard Optional(Dell USB Keyboard)-[Optional(16700)-Optional(8208)-Optional(Dell)-nil]
set input source to <TSMInputSource 0x7fe5bd211920> KB Layout: French (id=1) for keyboard Optional(Apple Internal Keyboard / Trackpad)-[Optional(1452)-Optional(628)-Optional(Apple Inc.)-Optional(D3H51914381FTV4AQ2BS)]
set input source to <TSMInputSource 0x7fe5bb70f320> KB Layout: Colemak (id=12825) for keyboard Optional(Dell USB Keyboard)-[Optional(16700)-Optional(8208)-Optional(Dell)-nil]
set input source to <TSMInputSource 0x7fe5bb503900> KB Layout: U.S. International - PC (id=15000) for keyboard Optional(Apple Internal Keyboard / Trackpad)-[Optional(1452)-Optional(628)-Optional(Apple Inc.)-Optional(D3H51914381FTV4AQ2BS)]
set input source to <TSMInputSource 0x7fe5bd2122e0> KB Layout: Colemak (id=12825) for keyboard Optional(Dell USB Keyboard)-[Optional(16700)-Optional(8208)-Optional(Dell)-nil]
set input source to <TSMInputSource 0x7fe5bd211680> KB Layout: U.S. International - PC (id=15000) for keyboard Optional(Apple Internal Keyboard / Trackpad)-[Optional(1452)-Optional(628)-Optional(Apple Inc.)-Optional(D3H51914381FTV4AQ2BS)]
```
Here I started with my internal keyboard, started typing on the dell external keyboard started typing on the internal keyboard again (sanity check). I changed the internal keyboard layout to Colemak. typed on the dell keyboard (it switched to french) changed the dell keyboard to US, switched  a couple times between the  internal keyboard (colemak was restored) and the dell keyboard (US was restored).

I'll add a trace at that level to notify that the layout associated with a given keyboard has been changed. 

in the meantime can you run a very quick test applying -v twice (so $path/autokbisw -v -v) the verbose flag is repeatable and this will set verbosity to level 2 which logs every event  received from every keyboard. once this runs simply type a couple keys on each keyboard. 
I am a bit worried that maybe karabiner is intercepting the events before autokbisw can get them.

**Comment by jeantil** [comment]
@ELLIOTTCABLE did you manage to get a trace ?

thanks

**Comment by belm0** [comment]
> set input source to <TSMInputSource 0x7f90a3f0dae0> KB Layout: U.S. (id=0) for keyboard Optional(**Karabiner VirtualHIDKeyboard**)-[Optional(5824)-Optional(10203)-Optional(pqrs.org)-Optional(pqrs.org:org_pqrs_driver_Karabiner_VirtualHIDDevice_VirtualHIDKeyboard_v060800)]

from what I've seen, autokbisw is not going to work in combination with Karabiner, since the latter channels all keyboards through a single virtual keyboard

**Comment by jeantil** [comment]
hmm I investigated karabiner before I wrote autokbisw. that was a couple years ago but I think I concluded that karabiner was able to do what autokbisw does and much much more. I wrote autokbisw because I didn't want to install such a complex software only for such a simple feature. 

best regards,
jean
